pkg/trie/triedb/node.go

```diff
@@ -0,0 +1,195 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package triedb
+
+import (
+	"bytes"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/pkg/trie/db"
+	"github.com/ChainSafe/gossamer/pkg/trie/triedb/codec"
+)
+
+type StorageHandle struct{ int }
+
+func (sh StorageHandle) toNodeHandle() NodeHandle {
+	return InMemory{idx: sh}
+}
+
+type NodeHandle interface {
+	isNodeHandle()
+}
+
+type (
+	InMemory struct {
+		idx StorageHandle
+	}
+	Hash struct {
+		hash common.Hash
+	}
+)
+
+func (InMemory) isNodeHandle() {}
+func (Hash) isNodeHandle()     {}
+
+func newFromEncodedMerkleValue(
+	parentHash common.Hash,
+	encodedNodeHandle codec.MerkleValue,
+	storage NodeStorage,
+) (NodeHandle, error) {
+	switch encoded := encodedNodeHandle.(type) {
+	case codec.HashedNode:
+		return Hash{hash: common.NewHash(encoded.Data)}, nil
+	case codec.InlineNode:
+		child, err := newNodeFromEncoded(parentHash, encoded.Data, storage)
+		if err != nil {
+			return nil, err
+		}
+		return InMemory{storage.alloc(New{child})}, nil
+	default:
+		panic("unreachable")
+	}
+}
+
+type Value interface {
+	isValue()
+	getHash() common.Hash
+}
+
+type (
+	Inline struct {
+		Data []byte
+	}
+
+	ValueRef struct {
+		hash common.Hash
+	}
+
+	NewValueRef struct {
+		hash *common.Hash
+		Data []byte
+	}
+)
+
+func (Inline) isValue()                  {}
+func (Inline) getHash() common.Hash      { return common.EmptyHash }
+func (ValueRef) isValue()                {}
+func (vr ValueRef) getHash() common.Hash { return vr.hash }
+func (NewValueRef) isValue()             {}
+func (vr NewValueRef) getHash() common.Hash {
+	if vr.hash == nil {
+		return common.EmptyHash
+	}
+
+	return *vr.hash
+}
+
+func NewValue(data []byte, threshold int) Value {
+	if len(data) >= threshold {
+		return NewValueRef{Data: data}
+	}
+
+	return Inline{Data: data}
+}
+
+func NewFromEncoded(encodedValue codec.NodeValue) Value {
+	switch encoded := encodedValue.(type) {
+	case codec.InlineValue:
+		return Inline{Data: encoded.Data}
+	case codec.HashedValue:
+		return ValueRef{hash: common.NewHash(encoded.Data)}
+	}
+
+	return nil
+}
+
+func InMemoryFetchedValue(value Value, prefix []byte, db db.DBGetter, fullKey []byte) ([]byte, error) {
```

Comment:
Does this mean the value being fetched is in memory? If so I'm a bit confused by the `db.Get()` call in the ValueRef case

Reply:
Yes, this method is used to get the value from a node that is in memory, in other words, we are holding nodes in memory before applying a commit to the db. 
If someone is looking up for that node before the commit is applied for example using `trieDB.lookup` we have to traverse these in-memory nodes and get the value.

There are 3 types of values:

- **Inline or NewValueRef:** this means that the value is part of the node and we can get it directly, but the difference will be more notable when we stored the values, since inline is stored inlined in the node and NewValueRef will create a new node that contains the value only (used in trie V1)
- **ValueRef:** is a reference to a value that is in another node in our DB so we have to execute a `db.Get` to get that value